### DISCUSSION TITLE
Is there any way to make FM faster?

### DISCUSSION THREAD

**Discussion by lorezyra** [estruyf/vscode-front-matter#429] [Q&A]
## Intro
I'm using HEXO.io for generating my Blog. I'm currently redesigning my [blog](https://2022.Blog.RichieBartlett.com). I have over 1400 posts (on production site) and will likely have beyond 5000 when I finish migrating from my old solution. 

## Hardware specs
- MacBookPro 16" (intel proc) 
- 64GB RAM
- high-end spec graphics
- 8TB storage. 

## Software specs
- MacOS 12.5.1
- VSCODE: 1.71.2 (universal)
- CHROME 105
- Safari 15.6.1
- HEXO 6.3
- nodeJS 18.9.1
- npm 8.19.1
- HomeBrew 3.6.2
- FM Version: 8.1.1

## FrontMatter
- Looks promising for easing my workflow for creating / managing my blog. However, it's a bit slow and cumbersome. It's obviously not designed with Hexo in mind. I understand that it's trying to remain a solution-for-all... but in its current form, it's a lot of effort to get it working with my setup for Hexo. 
- While I'll admit my ignorance on FM, the docs helped some, but there are a couple areas that could be improved. I've already submitted issues:
    - https://github.com/estruyf/vscode-front-matter/issues/428
    - https://github.com/estruyf/vscode-front-matter/issues/427 
    - https://github.com/estruyf/vscode-front-matter/issues/426 
    - https://github.com/estruyf/vscode-front-matter/issues/425
- In experimenting with FM for the past few days, I've found it looks nice but needs more work. Unfortunately, I have a lot on my plate already and don't see myself putting much more work into FM. As a developer, I'm comfortable with code and the console. So, it's not a big deal to operate without FM. However, I was thinking how I could make something others on my team could use. Enter: FM.
- I've found it a bit slow. I'm guessing it wasn't designed for someone with more than perhaps a 100 posts. I'm looking at my blog probably growing over 10K post in the next five years as I move _away_ from social media sites. At 1400+ posts, it often becomes non-responsive - forcing me to close the dashboard and restart it. 
- As you can see above, my hardware has no issues managing my blog. The design of this extension needs a bit more love and care than I can afford right now.

**Comment by estruyf** [comment]
Thanks, @lorezyra. Like any other open-source project, we all require feedback to improve things, and your feedback helps to understand your use case.

Also, this project is actively maintained by myself at the moment, and I cannot work on all the features at once. That is also why I sometimes need to understand better what you want to achieve to make the right decisions for the future of this product. For instance, when something might be broken, should it be fixed or should it be enhanced? A good example is a pagination issue you opened, this would be better to get enhanced.

> Looks promising for easing my workflow for creating / managing my blog. However, it's a bit slow and cumbersome. It's not designed with Hexo in mind. I understand that it's trying to remain a solution-for-all... but in its current form, it's a lot of effort to get it working with my setup for Hexo.

The original intention is to allow you to pick your SSG and use it in combination with FM. This is because some other CMSs have an opinion on what you should use or how. 

You are also the first Hexo user (apart from myself, who tried it out a long time ago). With your feedback, we'll improve the product to get better integration with Hexo but still have the flexibility to support other SSGs. The strength of FM should be in its configuration. Allow you to use it the way you defined it in your SSG. 

> While I'll admit my ignorance on FM, the docs helped some, but there are a couple of areas that could be improved. I've already submitted issues: ...

This is great feedback and comes back to me writing the documentation. Hopefully, with more FM users, the documentation will get better over time but highlighting things that are missing or need better clarity helps me understand what to do. Thank you for this.

> In experimenting with FM for the past few days, I've found it looks nice but needs more work. Unfortunately, I have a lot on my plate already and don't see myself putting much more work into FM. As a developer, I'm comfortable with code and the console. So, it's not a big deal to operate without FM. However, I was thinking how I could make something others on my team could use. Enter: FM.

I very much appreciate it you put the time and effort into trying out FM and providing your feedback. You are more than welcome to contribute to the project if you want.

> I've found it a bit slow. I'm guessing it wasn't designed for someone with more than perhaps a 100 posts. I'm looking at my blog probably growing over 10K post in the next five years as I move away from social media sites. At 1400+ posts, it often becomes non-responsive - forcing me to close the dashboard and restart it.

I always use my blog as a reference with almost 500 articles and over 1000's assets. There have already been a lot of enhancements for loading the experiences faster, but still, a lot can be done more/better.

We already have some caching and pagination in place, but still, what I want to have is process data in smaller batches. This will help a lot to load your project.

BTW: would you be able to share the project with me? It can be privately, of course, but I think it will be a good benchmarking project for future enhancements.

> As you can see above, my hardware has no issues managing my blog. The design of this extension needs a bit more love and care than I can afford right now.

Your hardware is not an issue 😁.

About UX and design, you are right, and there is an issue open on this matter: https://github.com/estruyf/vscode-front-matter/issues/94 - It would be great to get more people involved in it to make the product better. I believe we can do this with people like you that provide us valuable feedback 🙏.

**Comment by lorezyra** [comment]
Interesting. That would explain why the HEXO logo is on the site, but missing from the frameworks selection. I presume the FM adjusts how it handles pathway/folder-structure/etc with different frameworks. (I haven't dug into the extension code yet... no time now.)

As I mentioned, I have a growing blog with over 7900 files (assets and articles). It takes my MacBookPro about a minute to generate the site. I don't want to host any CMS solution for my blog - too many security risks. FM fits the bill nicely on that front. It's trivial to have everyone on my team connect to my GitHub repo from VScode and pull all the config and project files. Then they run a simple script to finish their setup. (That's the vision anyway...)

If FM could make it trivial for non-programmers to create blogs without performance issues or bugs, then it's a winner.

**Comment by estruyf** [comment]
> I presume the FM adjusts how it handles pathway/folder-structure/etc with different frameworks.

Not exactly, it seems that HEXO is the first one that approaches it differently, which would make it also the first one for which we'll adjust the behavior.

> That would explain why the HEXO logo is on the site, but missing from the frameworks selection.

The framework selection came over time, it was not there in the first versions, and each framework was added upon request, so that is the reason why it is not there yet.

> I don't want to host any CMS solution for my blog - too many security risks.

Also, a reason why it was created.

> If FM could make it trivial for non-programmers to create blogs without performance issues or bugs, then it's a winner.

Let us make it happen.
